src/mds/MDSRank.cc

```diff
@@ -1936,6 +1936,14 @@ bool MDSRankDispatcher::handle_asok_command(
       return true;
     }
     command_export_dir(f, path, (mds_rank_t)rank);
+  } else if (command == "dump inode") {
```

Comment:
This sounds useful but I would prefer this took an actual inode number rather than a path.

In any case, I think the `dump tree` command with a depth of 0 would be equivalent to your proposed `dump inode`, right?

Reply:
great point 👍 , didn't notice that dump tree command with depth == 0...

Regarding to the implementation of dump tree command, it would still go through all the inodes in cache, when the cache is too large, i think it's too expensive to use "dump tree" just for one inode, since  commands would hold the mds_lock. and i think the user may want to use "dump inode" frequently to monitor that specific inode sometime.

 so, can i do this by implementing two commands:
dump_inode_by_path <path>
dump_inode_by_ino  <inode number>

Reply:
I think `dump tree` only lists the given inode if you give depth `0` (or is it `1`?), right? The size of the cache wouldn't matter?

Reply:
yes, it only list the given inode, but the algorithm is traversing all the inodes in inode_map no matter if it's already found or not.  for huge cache and frequent use scenario,  i'm worried it would hold the mds_lock too long to just dump one inode.

Reply:
Okay... that `MDCache::dump_cache` code is a bit messy. After thinking about this more, I would like to put a band-aid on the `dump_tree` command where `MDCache::dump_cache` will specialize the case where `dump_root !- ""` and `depth == 0`. In that case, it just looks up the file as in your `MDCache::dump_inode` function but prints it out exactly as in `MDCache::dump_cache` (i.e. don't change the format of the output).

I also noticed that `MDCache::dump_cache` is dumping directories in a special way by dumping dirfrags whereas `CDir::dump` does not. Perhaps we can add a flags argument to `CDir::dump` similar to what you did for `CInode::dump`. So, `CDir::dump` would also dump its `dirs` array of dentries (but not by default). Simlarly, we would also have a flag for `CInode::dump` to dump its in-cache dirfrags.

In summary, we'd have the `dump_func` closure in `MDCache::dump_cache` just call `CInode::dump` with the correct flags after filtering the inodes by path.